sdk/core/Azure.Core/src/StatusCodeClassifier.cs

```diff
@@ -14,91 +12,56 @@ namespace Azure.Core
 
         /// <summary>
-        /// Creates a new instance of <see cref="StatusCodeClassifier"/>
+        /// Creates a new instance of <see cref="StatusCodeClassifier"/>.
         /// </summary>
-        /// <param name="successStatusCodes">The status codes that this classifier will consider
-        /// not to be errors.</param>
+        /// <param name="successStatusCodes">The status codes that this classifier
+        /// will consider not to be errors.</param>
         public StatusCodeClassifier(ReadOnlySpan<ushort> successStatusCodes)
         {
-            _successCodes = new ulong[Length];
+            _successCodes = new();
 
             foreach (int statusCode in successStatusCodes)
             {
                 AddClassifier(statusCode, isError: false);
             }
         }
 
-        private StatusCodeClassifier(ulong[] successCodes, ResponseClassificationHandler[]? handlers)
+        private StatusCodeClassifier(BitVector640 successCodes, ResponseClassificationHandler[]? handlers)
         {
-            Debug.Assert(successCodes?.Length == Length);
-
-            _successCodes = successCodes!;
+            _successCodes = successCodes;
             Handlers = handlers;
         }
 
         /// <inheritdoc/>
         public override bool IsErrorResponse(HttpMessage message)
         {
-            bool isError;
-
             if (Handlers != null)
             {
-                foreach (var handler in Handlers)
+                foreach (ResponseClassificationHandler handler in Handlers)
                 {
-                    if (handler.TryClassify(message, out isError))
+                    if (handler.TryClassify(message, out bool isError))
                     {
                         return isError;
                     }
                 }
             }
 
-            return !IsSuccessCode(message.Response.Status);
+            return !_successCodes[message.Response.Status];
         }
 
         internal virtual StatusCodeClassifier Clone()
-        {
-            ulong[] successCodes = new ulong[Length];
-            Array.Copy(_successCodes, successCodes, Length);
-
-            return new StatusCodeClassifier(successCodes, Handlers);
-        }
+            => new(_successCodes, Handlers);
 
         internal void AddClassifier(int statusCode, bool isError)
         {
             Argument.AssertInRange(statusCode, 0, 639, nameof(statusCode));
 
-            var index = statusCode >> 6;        // divides by 64
-            int bit = statusCode & 0b111111;    // keeps the bits up to 63
-            ulong mask = 1ul << bit;      // shifts a 1 to the position of code
-
-            ulong value = _successCodes[index];
-            if (!isError)
-            {
-                value |= mask;
-            }
-            else
-            {
-                value &= ~mask;
-            }
-
-            _successCodes[index] = value;
-        }
-
-        private bool IsSuccessCode(int statusCode)
```

Comment:
What kind of crazy magic is this?  I'm both fascinated and afraid of it.

Reply:
If the comments here are insufficient to answer your question, let me know - I 'm happy to unpack it more as needed: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/System.ClientModel/src/Internal/BitVector640.cs#L35-L62

Reply:
> What kind of crazy magic is this? I'm both fascinated and afraid of it.

branchless programming: https://en.algorithmica.org/hpc/pipelining/branchless/

Reply:
@KrzysztofCwalina, IIRC the switch statement makes it not 100% branchless - I read the IL a few months ago.  But I didn't spend the time to figure out how to do better for this case.